services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java

```diff
@@ -201,7 +203,22 @@ public Response processGrantRequest() {
 
         switch (action) {
             case AUTHORIZATION_CODE:
-                return codeToToken();
+                Object result = KeycloakModelUtils.runJobInRetriableTransaction(this.session.getKeycloakSessionFactory(), kcSession -> {
```

Comment:
It would be nice to have a comment here explaining why this block gets a retry, while the other one doesn't, and why the `WebApplicationException` receives a special treatment (suspected reason: "because WebApplicationException captures final business states that won't change when being retried like an invalid code").

Reply:
Agree.. added some comments that explain what is being done and why.